### PR TITLE
Improve flexibility of Slack connector

### PR DESCRIPTION
**Pull request by kearnsw**
**Proposed changes**:
By default:
- Bot responds when mentioned with @ symbol
- Bot responds when directly messaged

Previously, this functionality was either impossible or difficult for a new user to implement. See full description in #4811 

**Status (please check what you already did)**:
- [X] made PR ready for code review
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa_nlu#code-style) for instructions)

### CODE REVIEW

File: rasa/core/channels/slack.py [hunk 1]
```diff
@@ -319,18 +333,30 @@ def blueprint(self, on_new_message):
...
-                        ),
-                        sender_id=output.get("event").get("user"),
-                    )
+                    channel = output["event"]["channel"]
+                    if (
+                        is_direct_message(output)
+                        or is_app_mention(output)
+                        or channel == self.slack_channel
+                    ):
+                        self.set_output_channel(channel)
```
Comment (tmbo) [inline rasa/core/channels/slack.py]: we can't change the channel of the slack input. it is shared between all the async incoming messages so you might change it while another request is also getting processed (this problem might not surface right now, if the asyncs & awaits are at the right )

Comment (tmbo) [reply rasa/core/channels/slack.py]: I think we should remove that object state on the input channel and directly pass it to the output channel

File: rasa/core/channels/slack.py [hunk 2]
```diff
@@ -109,6 +109,20 @@ def _get_text_from_slack_buttons(buttons):
         return super(SlackBot, self).api_call("chat.postMessage", **json_message)
 
 
+def is_app_mention(user_message) -> bool:
```
Comment (tmbo) [inline rasa/core/channels/slack.py]: please add tests for these as well as type annotations for the parameters & a doc string

**Comment by erohmensing** [comment]
@kearnsw are you still planning to move forward with this PR?

**Comment by kearnsw** [comment]
@erohmensing: Yes, I am not intimately familiar with the way channels work so I need to spend some time reviewing that part of the codebase as @tmbo suggests.

File: rasa/core/channels/slack.py (hunk 2 above)
Comment (kearnsw) [reply rasa/core/channels/slack.py]: I've added the type annotations and standardized these with the other functions in the SlackInput channel. However, I'm not certain how to proceed with the test cases as the current channel tests for slack do not have a token. Is there a mock testing kit you'd prefer that I use for these?

File: rasa/core/channels/slack.py (hunk 1 above)
Comment (kearnsw) [reply rasa/core/channels/slack.py]: @tmbo , I am using the metadata object to store this value as well as others related to a specific user request. I think we'll need to explore incorporating the new converation session feature from 1.6.0.

**Comment by wochinge** [comment]
@kearnsw Let us know if you have any questions or need feedback :-)

**Comment by kearnsw** [comment]
> @kearnsw Let us know if you have any questions or need feedback :-)

There was a typing issue that slipped in when I made the prior updates. I've uploaded the latest after testing everything with pytype and formatted the code with black. Can you provide feedback on the updates? I have taken @tmbo's suggestion and am no longer setting the output channel property, and instead using the metadata parameter to store request specific information. The current implementation will leave the output channel configured static, while allowing for async messages based on the request from the sender.

File: rasa/core/channels/slack.py [hunk 3]
```diff
@@ -311,6 +329,18 @@ def _get_interactive_response(action: Dict) -> Optional[Text]:
 
         return response.text("")
 
+    def get_metadata(self, request: Request):
+        slack_event = request.json
+        event = slack_event.get("event")
+
+        metadata = {}
+        metadata["out_channel"] = event.get("channel")
```
Comment (wochinge) [inline rasa/core/channels/slack.py]: I just checked https://api.slack.com/types/event and there the `channel` param is part of `item`? Would also be awesome to add the link to https://api.slack.com/types/event as a comment in the code

File: rasa/core/channels/slack.py [hunk 4]
```diff
@@ -348,19 +378,32 @@ def blueprint(
...
-                    return await self.process_message(
-                        request,
-                        on_new_message,
-                        self._sanitize_user_message(
-                            output["event"]["text"], output["authed_users"]
-                        ),
-                        output.get("event").get("user"),
-                        metadata,
-                    )
+                    if (
```
Comment (wochinge) [inline rasa/core/channels/slack.py]: what do we need this condition for?

File: rasa/core/channels/slack.py [hunk 5]
```diff
@@ -311,6 +329,18 @@ def _get_interactive_response(action: Dict) -> Optional[Text]:
 
         return response.text("")
 
+    def get_metadata(self, request: Request):
```
Comment (wochinge) [inline rasa/core/channels/slack.py]: Can we please unit test this method?

**Comment by kearnsw** [comment]
@wochinge , can you advise on how to unit test the slack connector, since this requires API credentials. I do not see a precedent in the test suite currently. Thank you!

**Comment by wochinge** [comment]
@kearnsw Which methods do you want to unit test? I suggest targeting the methods individually so you don't have to deal with any actual API request (e.g. `is_app_mention` can be tested with a dictionary which you pass in). Does that make sense?

**Comment by kearnsw** [comment]
> @kearnsw Which methods do you want to unit test? I suggest targeting the methods individually so you don't have to deal with any actual API request (e.g. `is_app_mention` can be tested with a dictionary which you pass in). Does that make sense?

I can proceed that way. I wasn't sure if there would be interest in setting up the CI/CD pipeline with environment variables for testing the Slack interface, in case there are breaking changes to the Slack API. Thanks for the feedback, @wochinge!

File: rasa/core/channels/slack.py (hunk 4 above)
Comment (kearnsw) [reply rasa/core/channels/slack.py]: This is a holdover from the original codebase. I added the checks for direct message and app mention events. I think the response text should probably be set to "Received message on unsupported channel" or something to that effect instead of a blank response.

File: rasa/core/channels/slack.py (hunk 3 above)
Comment (kearnsw) [reply rasa/core/channels/slack.py]: That's odd. This is the event object that I receive from the slack connector: 

``` javascript
{ 'client_msg_id': 'XXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX, 
  'type': 'message', 
  'text': 'hi', 
  'user': 'XXXXXXXXX', 
  'ts': '1579802617.000800', 
  'team': 'XXXXXXXXX', 
  'blocks': [{'type': 'rich_text', 'block_id': 'XXXXX', 'elements': [{'type': 'rich_text_section', 'elements': [{'type': 'text', 'text': 'hi'}]}]}], 
  'channel': 'XXXXXXXXX', 
  'event_ts': '1579802617.000800', 
  'channel_type': 'im'
}
```

File: rasa/core/channels/slack.py [hunk 6]
```diff
@@ -348,19 +385,36 @@ def blueprint(
...
+                        )
 
             return response.text("Bot message delivered")
 
         return slack_webhook
 
-    def get_output_channel(self) -> OutputChannel:
-        return SlackBot(self.slack_token, self.slack_channel)
+    def get_output_channel(self, channel: Optional[Text] = None) -> OutputChannel:
+        if channel is not None:
```
Comment (wochinge) [inline rasa/core/channels/slack.py]: optional: I'd turn it around (`if is None else ...`) because it's a bit easier to read (you don't have to negate in your head), but that's really just a personal tase

File: tests/core/test_channels.py [hunk 7]
```diff
@@ -462,6 +462,50 @@ def test_botframework_attachments():
...
+
+    input_channel = SlackInput(
+        slack_token="YOUR_SLACK_TOKEN", slack_channel="YOUR_SLACK_CHANNEL"
+    )
+
+    r = Request(None, None, None, None, None, app="test")
+    r.parsed_json = direct_message_event
+    metadata = input_channel.get_metadata(request=r)
+    assert metadata["sender"] == user
+    assert metadata["out_channel"] == channel
```
Comment (wochinge) [inline tests/core/test_channels.py]: let's also test for the `authed_users` and for the case that an event doesn't contain some of these keys

File: tests/core/test_channels.py [hunk 8]
```diff
@@ -462,6 +462,50 @@ def test_botframework_attachments():
     assert ch.add_attachments_to_metadata(payload, metadata) == updated_metadata
 
 
+def test_slack_metadata():
```
Comment (wochinge) [inline tests/core/test_channels.py]: Thanks for adding the test 🚀

File: rasa/core/channels/slack.py (hunk 6 above)
Comment (wochinge) [inline rasa/core/channels/slack.py]: a short version would be 
```suggestion
        channel = channel or self.slack_channel
        return SlackBot(self.slack_token, channel)
```

File: rasa/core/channels/slack.py [hunk 9]
```diff
@@ -311,6 +329,25 @@ def _get_interactive_response(action: Dict) -> Optional[Text]:
...
         return response.text("")
 
+    def get_metadata(self, request: Request) -> Dict[Text, Any]:
+        """Extracts the metadata from a slack API event (https://api.slack.com/types/event).
+
+        Args:
+            request: a `Request` object that contains a slack API event in the body.
+
+        Returns:
+            A `dict` containing the output channel for the response, the text from the user, the sender's ID, and
```
[the remark this answers is not in the document]
Comment (wochinge) [reply rasa/core/channels/slack.py]: @kearnsw Have you seen that one?

File: rasa/core/channels/slack.py [hunk 10]
```diff
@@ -411,7 +411,8 @@ def blueprint(
         return slack_webhook
 
     def get_output_channel(self, channel: Optional[Text] = None) -> OutputChannel:
-        if channel is not None:
+        channel = channel or self.slack_channel
+        return SlackBot(self.slack_token, channel)
             return SlackBot(self.slack_token, channel)
```
Comment (wochinge) [inline rasa/core/channels/slack.py]: We should still delete the lines after

File: CHANGELOG.rst [hunk 11]
```diff
@@ -54,6 +54,11 @@ Features
   ``intent_tokenization_flag``: Flag to check whether to split intents (default ``False``).
   ``intent_split_symbol``: Symbol on which intent should be split (default ``_``)
 
+- `#4811 <https://github.com/RasaHQ/rasa/issues/4811>`_: Support invoking a ``SlackBot`` by direct messaging or @ mentions.
```
Comment (wochinge) [inline CHANGELOG.rst]: @kearnsw Please don't edit the `Changelog.rst` manually. Instead, create a file `4812.improvement.rst` in the `changelog` directory (the process is also described in here https://github.com/RasaHQ/rasa/blob/master/changelog/README.md )

File: docs/user-guide/connectors/slack.rst [hunk 12]
```diff
@@ -53,9 +53,9 @@ e.g. using:
 
 You need to supply a ``credentials.yml`` with the following content:
 
-- The ``slack_channel`` is the target your bot posts to.
-  This can be a channel or an individual person. You can leave out
-  the argument to post DMs to the bot.
+- The ``slack_channel`` is a channel that the bot will listen to for messages
+  in addition to direct messages and app mentions, i.e. "@app_name".
+  This can be a channel or an individual person.
```
Comment (wochinge) [inline docs/user-guide/connectors/slack.rst]: what do you mean by "This can be a channel or an individual person."?

File: tests/core/test_channels.py [hunk 13]
```diff
@@ -462,6 +462,94 @@ def test_botframework_attachments():
...
+            "event_ts": "1579802617.000800",
+            "channel_type": "im",
+        },
+    }
+
+    input_channel = SlackInput(
+        slack_token="YOUR_SLACK_TOKEN", slack_channel="YOUR_SLACK_CHANNEL"
+    )
+
+    r = Request(None, None, None, None, None, app="test")
```
Comment (wochinge) [inline tests/core/test_channels.py]: we recently discovered that this will lead to an error when we upgrade dependencies, let's rather do:
```suggestion
    r = Mock()
    r.json = direct_message_event
```

File: docs/user-guide/connectors/slack.rst (hunk 12 above)
Comment (kearnsw) [reply docs/user-guide/connectors/slack.rst]: @wochinge, this is from the original documentation. I added the part about "in addition to direct messages and app mentions, i.e. "@app_name". I'll try to clarify this sentence overall.

**Comment by kearnsw** [comment]
The build is failing on tests unrelated to the modified code. Can you advise @wochinge ?

**Comment by wochinge** [comment]
Hi @kearnsw ,

can you please rebase to / merge in master ? We switched to GitHub actions which are more reliable and also faster.

File: CHANGELOG.rst [hunk 14]
```diff
@@ -1,8 +1,8 @@
-:desc: Rasa Open Source Changelog
+:desc: Rasa OSS Changelog
```
Comment (wochinge) [inline CHANGELOG.rst]: can you please undo this change?

File: CHANGELOG.rst [hunk 15]
```diff
@@ -1,8 +1,8 @@
...
 
 
-Rasa Open Source Change Log
-===========================
+Rasa OSS Change Log
```
Comment (wochinge) [inline CHANGELOG.rst]: can you please undo this change?

File: CHANGELOG.rst [hunk 16]
```diff
@@ -17,89 +17,6 @@ This project adheres to `Semantic Versioning`_ starting with version 1.0.
 
 .. towncrier release notes start
 
-[1.7.0] - 2020-01-29
```
Comment (wochinge) [inline CHANGELOG.rst]: can you please undo this change?

File: changelog/4811.improvement.rst [hunk 17]
```diff
@@ -0,0 +1 @@
+Support invoking a ``SlackBot`` by direct messaging or @ mentions.
```
Comment (wochinge) [inline changelog/4811.improvement.rst]:
```suggestion
Support invoking a ``SlackBot`` by direct messaging or ``@<app name>`` mentions.
```

File: rasa/core/channels/slack.py [hunk 18]
```diff
@@ -293,11 +307,15 @@ def _get_interactive_response(action: Dict) -> Optional[Text]:
 
             return response.text(None, status=201, headers={"X-Slack-No-Retry": 1})
 
+        if metadata is not None:
```
Comment (wochinge) [inline rasa/core/channels/slack.py]: this can't be `None` in your implementation, so you don't have to check for it.

File: rasa/core/channels/slack.py [hunk 19]
```diff
@@ -310,6 +328,25 @@ def _get_interactive_response(action: Dict) -> Optional[Text]:
...
+
+        Returns:
+            Metadata extracted from the sent event payload. This includes the output channel for the response, the text
+            from the user and users that have installed the bot.
+        """
+        slack_event = request.json
+        event = slack_event.get("event", {})
+        return {
+            "out_channel": event.get("channel"),
+            "text": event.get("text"),
```
Comment (wochinge) [inline rasa/core/channels/slack.py]: it feels a bit weird that the `metadata` also returns `sender` and `text`. The metadata is meant to provide additional information around a message. `text` and `sender` are the message. So in my opinion it would be a bit cleaner to keep this separated, what you think?

File: rasa/core/channels/slack.py (hunk 18 above)
Comment (kearnsw) [reply rasa/core/channels/slack.py]: This is required for typing. As process_message, contains metadata as an option.

**Comment by kearnsw** [comment]
@wochinge, looks like the build environment errored out.

File: CHANGELOG.rst [hunk 20]
```diff
@@ -133,6 +133,7 @@ Bugfixes
   to the updated doc link.
 
 
+>>>>>>> 5a3b9d254164a0f17892bb85747de2a73a9a0c63
```
Comment (wochinge) [inline CHANGELOG.rst]:
```suggestion
```

**Comment by wochinge** [comment]
@kearnsw Can you please update with the latest master again?

**Comment by wochinge** [comment]
@kearnsw I'm sorry, but you have to merge in the latest master state again 🙈

**Comment by kearnsw** [comment]
Hi @wochinge , I’m not understanding the errors for the past few merge builds. Are these backend issues or problems with the submitted code?

**Comment by tmbo** [comment]
Looks all good :+1:
